### PR TITLE
#42803 Save_to_path function now supports formats other than psd

### PR DESCRIPTION
**Pull request by hawif**
Added support for different file formats to the save_to_path function.

### CODE REVIEW

File: engine.py [hunk 1]
```diff
@@ -504,9 +504,53 @@ def save_to_path(self, document, path):
...
+                                 ".tga" : self.adobe.TargaSaveOptions,
+                                 ".targa" : self.adobe.TargaSaveOptions,
+
+                                 ".tif" : self.adobe.TiffSaveOptions,
+                                 ".tiff" : self.adobe.TiffSaveOptions
+ 
+                                 }
+           
+            # if extension not in list, save file as .psd
+            if ext.lower() not in valid_extensions:
```
Comment (josh-t) [inline engine.py]: minor: I'd switch the conditional and remove `not` here. I think it will read better.

File: engine.py [hunk 2]
```diff
@@ -504,9 +504,53 @@ def save_to_path(self, document, path):
...
+                 save_options = self.adobe.PhotoshopSaveOptions
+            else:
+                 save_options = valid_extensions[ext.lower()]
+           
+            document.saveAs(self.adobe.File(path),save_options)
```
Comment (josh-t) [inline engine.py]: space after the comma

File: engine.py [hunk 3]
```diff
@@ -504,9 +504,53 @@ def save_to_path(self, document, path):
 
             # make the document being processed the active document
             self.adobe.app.activeDocument = document
-
-            # TODO need to use appropriate save options here (jpg, tiff, etc)
-            document.saveAs(self.adobe.File(path))
+        
+            (_, ext) = os.path.splitext(path)
+
+            # valid file extensions and respective save option in photoshop
```
Comment (josh-t) [inline engine.py]: Do any of these options require and non-default settings? I know we set this in the version upload code: https://github.com/shotgunsoftware/tk-photoshopcc/blob/master/hooks/tk-multi-publish2/basic/upload_version.py#L222

Not saying we need to do anything different here, just curious if the default options are sufficient in all cases.

File: engine.py [hunk 4]
```diff
@@ -504,9 +504,53 @@ def save_to_path(self, document, path):
...
-
-            # TODO need to use appropriate save options here (jpg, tiff, etc)
-            document.saveAs(self.adobe.File(path))
+        
+            (_, ext) = os.path.splitext(path)
+
+            # valid file extensions and respective save option in photoshop 
+            valid_extensions = {
+                                 ".bmp" : self.adobe.BMPSaveOptions, 
+                                 # DCS1_saveOptions and DCS2_SaveOptions skipped
```
Comment (josh-t) [inline engine.py]: maybe move all the "skipped" comments to the comment above the declaration with an explanation as to why we're skipping some.

File: engine.py [hunk 5]
```diff
@@ -504,9 +504,53 @@ def save_to_path(self, document, path):
...
+
+                                 ".pict" : self.adobe.PICTFileSaveOptions,
+                                 ".pct" : self.adobe.PICTFileSaveOptions,
+                                 ".pic" : self.adobe.PICTFileSaveOptions,
+                                 # PICTResourceSaveOptions skipped
+
+                                 ".pixar" : self.adobe.PixarSaveOptions,
+                                 ".png" : self.adobe.PNGSaveOptions,
+                                 ".raw"  : self.adobe.RawSaveOptions,
+
```
Comment (josh-t) [inline engine.py]: Any reason why there are spaces between these?

File: engine.py [hunk 6]
```diff
@@ -504,9 +504,53 @@ def save_to_path(self, document, path):
...
+                                 ".pdf" : self.adobe.PDFSaveOptions,
+                                 ".psd" : self.adobe.PhotoshopSaveOptions,
+
+                                 ".pict" : self.adobe.PICTFileSaveOptions,
+                                 ".pct" : self.adobe.PICTFileSaveOptions,
+                                 ".pic" : self.adobe.PICTFileSaveOptions,
+                                 # PICTResourceSaveOptions skipped
+
+                                 ".pixar" : self.adobe.PixarSaveOptions,
```
Comment (josh-t) [inline engine.py]: Looks like some of these are grouped by save options. I suggest either alphabetizing by the extension with no spaces.

File: engine.py (hunk 3 above)
Comment (hawif) [reply engine.py]: You are right, we need to specify the quality (default value is 3).

File: engine.py [hunk 7]
```diff
@@ -504,9 +504,48 @@ def save_to_path(self, document, path):
...
             self.adobe.app.activeDocument = document
-
-            # TODO need to use appropriate save options here (jpg, tiff, etc)
-            document.saveAs(self.adobe.File(path))
+        
+            (_, ext) = os.path.splitext(path)
+            ext = ext.lower()
+
+            # these extensions are supported by photoshop
+            # PICTResourceSaveOptions is skipped for now, need a way to differentiate PICT
```
Comment (josh-t) [inline engine.py]: Now that there are conditionals for each type, I'd suggest putting the comments inside the appropriate condition.

File: engine.py [hunk 8]
```diff
@@ -504,9 +504,48 @@ def save_to_path(self, document, path):
...
+                save_options = self.adobe.BMPSaveOptions
+            elif ext ==".dcs":
+                save_options = self.adobe.DCS2_SaveOptions
+            elif ext ==".eps":
+                save_options = self.adobe.EPSSaveOptions
+            elif ext ==".gif":
+                save_options = self.adobe.GIFSaveOptions
+            elif ext in [".jpg", ".jpeg"]:
+                save_options = self.adobe.JPEGSaveOptions;
+                save_options.quality = 12
```
Comment (josh-t) [inline engine.py]: Suggest a comment saying why we're setting this to something other than the default.

**Pull request merged**